feat: switch region during login and signup

kappa90 opened the pull request:
## Problem

We need to allow logging in and signing up from different regions, as we're about to launch Posthog EU.

## Changes

- Add a `RegionSelect` selector in the login and signup pages, which allows you to switch to a different region, keeping the same path
- Added a `REGION` env var, default set to `US`, which allows us to define the region for a deployment
- Changed the error message for invalid logins, to nudge users to check if they're logging in the right region
- Note: it's currently behind a feature flag, and only enabled for Cloud (as it should)
- Note: currently, hardcoded the urls for US and EU instances

<img width="925" alt="image" src="https://user-images.githubusercontent.com/7335343/191919917-33f41878-5817-4fb1-a018-b803a00289ab.png">
<img width="925" alt="image" src="https://user-images.githubusercontent.com/7335343/191921612-30316829-6ab5-428a-9cb9-45b24de18f7b.png">

## How did you test this code?
- Tested locally

lharries commented:
Good work! Confirming that the dropdown will be hidden for selfhosted deployments and won't cause any funky business?

Twixes commented on frontend/src/lib/constants.tsx:
```diff
@@ -185,3 +185,6 @@ export const UPGRADE_LINK = (cloud?: boolean): { url: string; target?: '_blank'
 export const DOMAIN_REGEX = /^([a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,}$/
 export const SECURE_URL_REGEX = /^(?:http(s)?:\/\/)[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:\/?#[\]@!\$&'\(\)\*\+,;=.]+$/gi
+
+export const CLOUD_US_URL = 'https://app.posthog.com'
+export const CLOUD_EU_URL = 'https://eu.posthog.com'
```
Nit but this could be just the hostname, as we for sure want `https://` always :D
Hardcoding overall seem pretty good as a solution – we're definitely not gonna need service discovery for regions. Suggestion though: this can be more obvious as a `Record<Region, string>` constant rather than separate constants

Twixes commented on frontend/src/scenes/authentication/RegionSelect.tsx:
```diff
@@ -0,0 +1,56 @@
+                return
+            }
+            const { pathname } = router.values.currentLocation
+            const newUrl = region === Region.EU ? `${CLOUD_EU_URL}${pathname}` : `${CLOUD_US_URL}${pathname}`
```
Could also include `search` and `hash` just in case (we might want to preserve UTM params or something like that)

Twixes commented on frontend/src/scenes/authentication/RegionSelect.tsx:
```diff
@@ -0,0 +1,56 @@
+import type { regionSelectLogicType } from './RegionSelectType'
+import { PureField } from 'lib/forms/Field'
+
+export const regionSelectLogic = kea<regionSelectLogicType>([
```
I'm all for logics but honestly at this level of complexity I think the `onRegionChange` handler could just be inlined in `RegionSelect`

kappa90 replied:
I was thinking on expanding the logic, but then I didn't 😅  I removed it now!

kappa90 commented:
> Good work! And love the emoji touch.
> 
> Confirming that the dropdown will be hidden for selfhosted deployments and won't cause any funky business?

It's only for cloud and behind a feature flag.

kappa90 commented on posthog/api/authentication.py:
```diff
@@ -86,7 +86,10 @@ def create(self, validated_data: Dict[str, str]) -> Any:
         if not user:
-            raise serializers.ValidationError("Invalid email or password.", code="invalid_credentials")
+            raise serializers.ValidationError(
+                "Invalid email or password. Are you trying to login from the correct region?",
```
Nice, couldn't find the right wording.

kappa90 commented on posthog/utils.py:
```diff
@@ -722,6 +722,15 @@ def get_instance_realm() -> str:
         return "hosted-clickhouse"
 
 
+def get_instance_region() -> str:
```
🆗

Twixes commented:
From an experience perspective, I'm a bit worried that putting the "Region" selector front and center might dissuade users from outside the US or EU to sign up.

Ways this is handled by other companies in a similar situation:
1. Zoho infers region based on location (country resolution), it's a small note at the bottom of the form. It's so implicit though, that as their user I was pretty surprised when I couldn't log into zoho.com. I just wasn't aware I was actually signed up for zoho.eu! – https://www.zoho.com/signup.html
2. Mixpanel made region an option in project creation (continent resolution), which seems super practical, though undoubtedly significantly more complex to implement. – https://help.mixpanel.com/hc/en-us/articles/360039135652-Data-Residency-in-EU
3. Amplitude has an EU hosting option too, though it seems somewhat obscure, I'm not actually sure if it's possible to get to it from the homepage. They do have a country selector in signup, but it's not clear whether that affects where the data is hosted. – https://amplitude.com/get-started-eu

The per-project (or per-organization) approach seems like the smoothest experience, but since that's a much different scope of work than currently defined for the EU hosting project.

So, as a simple change, maybe instead of saying:

**Region**
PostHog US

We could say:

**Data Region**
United States

What do you think?

lharries commented:
> **Data Region** United States
> 
> What do you think?

Strong +1 for Data Region to reduce risk of that

kappa90 committed:
feat: switch region during login and signup

kappa90 commented:
I updated the label to `Data region` and changed the options to `United States` and `European Union`.

I also cherry picked the error message changes and put them in a different PR -> https://github.com/PostHog/posthog/pull/11949

Why: because putting that change behind a feature flag would be a mess, since it's a backend setting. I'd rather just merge that change after Cloud EU has been released.

benjackwhite reviewed (approved):
Niiiice I like this a lot. Simple but effective.

lharries reviewed (approved):
Great work Em :)

Twixes reviewed (approved):
🚀

kappa90 committed:
add correct error message

kappa90 commented:
Reinstated the error message here, behind a feature flag 💪

kappa90 committed:
add correct error message

corywatilo commented:
Personal opinion: Sort of like saying `US (Virginia)` and `EU (Frankfurt)` because:

- it feels a little more like AWS, hence more control over data locality
- alludes to the fact we might have more regions in the future

"United States" and "European Union" feel very broad, and I think the above would connect well with our audience.

kappa90 commented:
> Personal opinion: Sort of like saying `US (Virginia)` and `EU (Frankfurt)` because:
> 
> * it feels a little more like AWS, hence more control over data locality
> * alludes to the fact we might have more regions in the future
> 
> "United States" and "European Union" feel very broad, and I think the above would connect well with our audience.
> 
> Also having a _More info_ link across from the label (or a (i) button) could offer a tooltip that explains the differences. (Tooltip so we don't have people leaving the page entirely)

Adding this suggestions in this PR that Joe started: https://github.com/PostHog/posthog/pull/12025

mariusandra commented on frontend/src/scenes/authentication/Login.tsx:
```diff
@@ -15,6 +15,7 @@ import { Form } from 'kea-forms'
 import { Field } from 'lib/forms/Field'
 import { AlertMessage } from 'lib/components/AlertMessage'
 import { BridgePage } from 'lib/components/BridgePage/BridgePage'
+import RegionSelect from './RegionSelect'
```
Let's avoid `default` exports whenever possible. https://posthog.com/handbook/engineering/conventions/frontend-coding#general-tips

mariusandra reviewed (commented):
Just one late nit
